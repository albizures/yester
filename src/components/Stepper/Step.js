import { Dimensions, View, StyleSheet } from 'react-native'
import PropTypes from 'prop-types'
import React from 'react'

const Step = (props) => {
  const { children } = props
  const { width, height } = Dimensions.get('window')
  const portrait = height > width
  const style = {
    justifyContent: portrait ? 'flex-start' : 'flex-start',
    paddingTop: portrait ? 50 : 10,
    width,
    height,
  }

  return (
    <View style={[styles.container, style]}>
      {children}
    </View>
  )
}

Step.propTypes = {
  children: PropTypes.node.isRequired,
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
  },
})

export default Step
