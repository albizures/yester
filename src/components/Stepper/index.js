import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { FlatList, StyleSheet, View } from 'react-native'

import Step from './Step'

export default class Stepper extends Component {
  static propTypes = {
    scrollEnabled: PropTypes.bool,
    children: PropTypes.arrayOf(PropTypes.node).isRequired,
    bottomBar: PropTypes.func,
  }

  flatList = React.createRef()

  state = {
    currentStep: 0,
    previousStep: null,
  }

  onSwipePageChange = ({ viewableItems }) => {
    if (!viewableItems[0] || this.state.currentStep === viewableItems[0].index) {
      return
    }

    this.setState(state => ({
      previousStep: state.currentStep,
      currentStep: viewableItems[0].index,
    }))
  }

  next = () => {
    const { current: flatList } = this.flatList
    const { children } = this.props
    const { currentStep } = this.state

    if (currentStep >= children.length - 1) {
      return
    }

    flatList.scrollToIndex({
      animated: true,
      index: currentStep + 1,
    })
  }

  skip = () => {
    const { current: flatList } = this.flatList
    const { children } = this.props

    flatList.scrollToIndex({
      animated: false,
      index: children.length - 1,
    })
  }

  keyExtractor = (item, index) => index.toString();

  renderItem = ({ item, index }) => (
    <Step index={index}>
      {React.cloneElement(item, {
        ...this.state,
        index,
        skip: this.skip,
        next: this.next,
      })}
    </Step>
  )

  render () {
    const { children, bottomBar, scrollEnabled } = this.props

    return (
      <View style={styles.container}>
        <FlatList
          ref={this.flatList}
          showsHorizontalScrollIndicator={false}
          horizontal
          pagingEnabled
          onViewableItemsChanged={this.onSwipePageChange}
          data={children}
          renderItem={this.renderItem}
          keyExtractor={this.keyExtractor}
          scrollEnabled={scrollEnabled}
        />
        {bottomBar && bottomBar({
          skip: this.skip,
          next: this.next,
          ...this.state,
        })}
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    width: '100%',
  },
})
