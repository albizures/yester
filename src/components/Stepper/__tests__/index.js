import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import Stepper from '../';
const Step = () => <div>step</div>;

const children = [<Step key={0} />, <Step key={1} />, <Step key={2} />];

const render = () => <Stepper>{children}</Stepper>;

describe('src/screens/Stepper', () => {
	it('should render', () => {
		const tree = renderer.create(render()).toJSON();
		expect(tree).toMatchSnapshot();
	});

	describe('when next is executed', () => {
		it('should scroll to the next step', () => {
			const shallowStepper = shallow(render());
			const instance = shallowStepper.instance();

			instance.flatList.current = {
				scrollToIndex: jest.fn(),
			};

			instance.next();

			expect(instance.flatList.current.scrollToIndex).toHaveBeenCalledTimes(1);
			expect(instance.flatList.current.scrollToIndex).toHaveBeenCalledWith({
				animated: true,
				index: 1,
			});
		});
		describe('when the current step is the last one', () => {
			it('shouldnÂ´t do nothing', () => {
				const shallowStepper = shallow(render());
				const instance = shallowStepper.instance();

				instance.state = {
					currentStep: 2,
					previousStep: 1,
				};
				instance.flatList.current = {
					scrollToIndex: jest.fn(),
				};

				instance.next();

				expect(instance.flatList.current.scrollToIndex).toHaveBeenCalledTimes(
					0,
				);
			});
		});
	});

	describe('when skip is executed', () => {
		it('should scroll to last step', () => {
			const shallowStepper = shallow(render());
			const instance = shallowStepper.instance();

			instance.flatList.current = {
				scrollToIndex: jest.fn(),
			};

			instance.skip();

			expect(instance.flatList.current.scrollToIndex).toHaveBeenCalledTimes(1);
			expect(instance.flatList.current.scrollToIndex).toHaveBeenCalledWith({
				animated: false,
				index: children.length - 1,
			});
		});
	});
	describe('when onSwipePageChange is executed', () => {
		it('should update the state', () => {
			const shallowStepper = shallow(render());
			const instance = shallowStepper.instance();

			expect(shallowStepper.state('currentStep')).toBe(0);
			expect(shallowStepper.state('previousStep')).toBe(null);

			instance.onSwipePageChange({
				viewableItems: [{ index: 1 }],
			});

			expect(shallowStepper.state('currentStep')).toBe(1);
			expect(shallowStepper.state('previousStep')).toBe(0);
		});

		describe('when there is no viewable items', () => {
			it('should not change the state', () => {
				const shallowStepper = shallow(render());
				const instance = shallowStepper.instance();

				expect(shallowStepper.state('currentStep')).toBe(0);
				expect(shallowStepper.state('previousStep')).toBe(null);

				instance.onSwipePageChange({
					viewableItems: [],
				});

				expect(shallowStepper.state('currentStep')).toBe(0);
				expect(shallowStepper.state('previousStep')).toBe(null);
			});
		});

		describe('when the viewable items are the same', () => {
			it('should not change the state', () => {
				const shallowStepper = shallow(render());
				const instance = shallowStepper.instance();

				expect(shallowStepper.state('currentStep')).toBe(0);
				expect(shallowStepper.state('previousStep')).toBe(null);

				instance.onSwipePageChange({
					viewableItems: [{ index: 0 }],
				});

				expect(shallowStepper.state('currentStep')).toBe(0);
				expect(shallowStepper.state('previousStep')).toBe(null);
			});
		});
	});
});
