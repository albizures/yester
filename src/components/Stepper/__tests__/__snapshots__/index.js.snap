// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/screens/Stepper should render 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "justifyContent": "flex-start",
    }
  }
>
  <RCTScrollView
    data={
      Array [
        <Step />,
        <Step />,
        <Step />,
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={true}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    onViewableItemsChanged={[Function]}
    pagingEnabled={true}
    renderItem={[Function]}
    scrollEventThrottle={50}
    showsHorizontalScrollIndicator={false}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={
      Array [
        Object {
          "onViewableItemsChanged": [Function],
          "viewabilityConfig": undefined,
        },
      ]
    }
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "flex": 1,
              },
              Object {
                "height": 1334,
                "justifyContent": "flex-start",
                "paddingTop": 50,
                "width": 750,
              },
            ]
          }
        >
          <div>
            step
          </div>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "flex": 1,
              },
              Object {
                "height": 1334,
                "justifyContent": "flex-start",
                "paddingTop": 50,
                "width": 750,
              },
            ]
          }
        >
          <div>
            step
          </div>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "flex": 1,
              },
              Object {
                "height": 1334,
                "justifyContent": "flex-start",
                "paddingTop": 50,
                "width": 750,
              },
            ]
          }
        >
          <div>
            step
          </div>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
